File: backend/autotool.py
"""OpenAPI tool acquisition and intent retrieval over short tool descriptions."""
from copy import deepcopy
import hashlib
import json
import re
import math
from collections import Counter
from urllib.parse import quote
from .tools import Tool, object_schema

UNSAFE = {'__proto__', 'constructor', 'prototype'}


def intent_tokens(text):
    words = re.findall(r'[a-z0-9]+', text.lower())
    for phrase in re.findall(r'[\u4e00-\u9fff]+', text):
        words.extend(phrase[i:i+2] for i in range(max(0, len(phrase)-1)))
    return words


def retrieve_tools(intent, tools, k=4):
    """Local BM25 over concise tool descriptions, not a semantic embedding model."""
    docs = [Counter(intent_tokens(t.name.replace('_', ' ') + ' ' + t.description)) for t in tools]
    query = set(intent_tokens(intent))
    mean = sum(sum(d.values()) for d in docs) / max(1, len(docs))
    df = Counter(word for doc in docs for word in doc)
    ranked = []
    for tool, doc in zip(tools, docs):
        score = 0.0
        for term in query:
            freq = doc[term]
            if freq:
                idf = math.log(1 + (len(docs) - df[term] + .5) / (df[term] + .5))
                score += idf * freq * 2.2 / (freq + 1.2 * (.25 + .75 * sum(doc.values()) / max(mean, 1)))
        ranked.append(dict(name=tool.name, score=round(score, 5)))
    return sorted(ranked, key=lambda r: (-r['score'], r['name']))[:k]


def canonical(value):
    return json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(',', ':'), allow_nan=False)


def digest(value):
    return hashlib.sha256(canonical(value).encode()).hexdigest()


def path_value(value, path):
    for part in path:
        if part in UNSAFE:
            raise ValueError('Unsafe graph path')
        if isinstance(value, dict) and part in value:
            value = value[part]
        elif isinstance(value, list) and str(part).isdigit() and int(part) < len(value):
            value = value[int(part)]
        else:
            raise ValueError('图参数路径不存在：' + '.'.join(path))
    return value


def acquire_tools(spec, transport, observe):
    if spec.get('openapi') != '3.0.3' or not spec.get('info', {}).get('title'):
        raise ValueError('Unsupported OpenAPI specification')
    result, names = [], set()
    for path, methods in spec['paths'].items():
        if not path.startswith('/api/') or '..' in path or '?' in path or set(methods) != {'get'}:
            raise ValueError('AutoTool only permits configured GET API paths')
        op = methods['get']
        name = op['operationId']
        if not re.fullmatch('[a-z][a-z0-9_]{0,63}', name) or name in names:
            raise ValueError('Duplicate or invalid operation ID')
        names.add(name)
        properties, required = {}, []
        for parameter in op['parameters']:
            key, schema = parameter['name'], deepcopy(parameter['schema'])
            if key in UNSAFE or key in properties or not re.fullmatch('[A-Za-z][A-Za-z0-9_]{0,63}', key) or parameter['in'] not in ['path', 'query'] or parameter['required'] is not True:
                raise ValueError('Unsupported or unsafe OpenAPI parameter')
            if schema['type'] not in ['string', 'integer', 'number', 'boolean']:
                raise ValueError('Unsupported OpenAPI parameter type')
            if schema['type'] == 'string':
                schema.setdefault('minLength', 1)
                schema.setdefault('maxLength', 500)
            properties[key] = schema
            required.append(key)
        placeholders = set(re.findall(r'\{([^}]+)\}', path))
        if placeholders != {p['name'] for p in op['parameters'] if p['in'] == 'path'}:
            raise ValueError('OpenAPI path parameters do not match placeholders')

        async def handler(args, context, path=path, op=op):
            query = {}
            for parameter in op['parameters']:
                value = args[parameter['name']]
                value = str(value).lower() if isinstance(value, bool) else str(value)
                if parameter['in'] == 'path':
                    if value in ['.', '..']:
                        raise ValueError('Dot segments are not valid record identifiers')
                    path = path.replace('{' + parameter['name'] + '}', quote(value, safe=''))
                else:
                    query[parameter['name']] = value
            response = await transport(path, query)
            data = path_value(response, op.get('x-data-path', []))
            return observe(data, op['x-resource'], context)

        result.append(Tool(name, op['summary'], 'read', object_schema(properties, required), handler,
                           {'kind': 'autotool', 'spec': spec['info']['title'], 'operationId': name, 'digest': digest(spec)}))
    return result

File: server/autotool.ts
import { z } from 'zod';
import { defineTool, type Tool, type ToolContext } from './tools.js';
import { digest } from './graph.js';

type Schema = { type: 'string' | 'integer' | 'number' | 'boolean'; enum?: (string | number)[]; minimum?: number; maximum?: number; minLength?: number; maxLength?: number; description?: string };
export interface ApiSpec {
  openapi: '3.0.3'; info: { title: string; version: string };
  paths: Record<string, { get: { operationId: string; summary: string; parameters: { name: string; in: 'path' | 'query'; required: true; schema: Schema }[]; 'x-resource': string; 'x-data-path'?: string[] } }>;
}
export type ApiTransport = (path: string, query: Record<string, string>, signal: AbortSignal) => Promise<unknown>;

/** Restricted OpenAPI compiler: generates actual typed HTTP tools, never evals code.
 * Specs are curated local API contracts; server URLs/credentials are never taken from specs.
 */
export function acquireTools(spec: ApiSpec, transport: ApiTransport, observe: (value: unknown, resource: string, context: ToolContext) => unknown): Tool[] {
  if (spec.openapi !== '3.0.3' || !spec.info?.title) throw new Error('Unsupported OpenAPI specification');
  const names = new Set<string>();
  return Object.entries(spec.paths).map(([path, methods]) => {
    if (!path.startsWith('/api/') || path.includes('..') || path.includes('?') || Object.keys(methods).some(key => key !== 'get')) throw new Error('AutoTool only permits configured GET API paths');
    const operation = methods.get;
    if (!/^[a-z][a-z0-9_]{0,63}$/.test(operation.operationId) || names.has(operation.operationId)) throw new Error('Duplicate or invalid operation ID');
    names.add(operation.operationId);
    const shape: z.ZodRawShape = {};
    for (const parameter of operation.parameters) {
      if (!/^[a-zA-Z][a-zA-Z0-9_]{0,63}$/.test(parameter.name) || ['constructor', 'prototype', '__proto__'].includes(parameter.name) || Object.hasOwn(shape, parameter.name) || !['path', 'query'].includes(parameter.in) || parameter.required !== true) throw new Error('Unsupported or unsafe OpenAPI parameter');
      const schema = parameter.schema;
      let validator: z.ZodTypeAny;
      if (schema.type === 'string') validator = z.string().min(schema.minLength ?? 1).max(schema.maxLength ?? 500);
      else if (schema.type === 'integer') validator = z.number().int().min(schema.minimum ?? 0).max(schema.maximum ?? Number.MAX_SAFE_INTEGER);
      else if (schema.type === 'number') validator = z.number().finite().min(schema.minimum ?? -Number.MAX_SAFE_INTEGER).max(schema.maximum ?? Number.MAX_SAFE_INTEGER);
      else if (schema.type === 'boolean') validator = z.boolean();
      else throw new Error('Unsupported OpenAPI parameter type');
      if (schema.enum) {
        if (!schema.enum.length) throw new Error('OpenAPI enum cannot be empty');
        const literals = schema.enum.map(value => z.literal(value));
        validator = validator.and(literals.length === 1 ? literals[0] : z.union(literals as [z.ZodLiteral<any>, z.ZodLiteral<any>, ...z.ZodLiteral<any>[]]));
      }
      shape[parameter.name] = schema.description ? validator.describe(schema.description) : validator;
    }
    const placeholders = [...path.matchAll(/\{([^}]+)\}/g)].map(match => match[1]);
    if (placeholders.some(name => !operation.parameters.some(parameter => parameter.in === 'path' && parameter.name === name)) || operation.parameters.some(parameter => parameter.in === 'path' && !placeholders.includes(parameter.name))) throw new Error('OpenAPI path parameters do not match placeholders');
    const tool = defineTool(operation.operationId, operation.summary, 'read', shape, async (args, context) => {
      let resolved = path;
      const query: Record<string, string> = {};
      for (const parameter of operation.parameters) {
        if (parameter.in === 'path') {
          const value = String(args[parameter.name]);
          if (value === '.' || value === '..') throw new Error('Dot segments are not valid record identifiers');
          resolved = resolved.replaceAll('{' + parameter.name + '}', encodeURIComponent(value));
        } else query[parameter.name] = String(args[parameter.name]);
      }
      let value = await transport(resolved, query, context.signal);
      for (const key of operation['x-data-path'] || []) {
        if (['__proto__', 'constructor', 'prototype'].includes(key) || !value || typeof value !== 'object' || !Object.hasOwn(value, key)) throw new Error('API response does not match the acquired data path');
        value = (value as Record<string, unknown>)[key];
      }
      return observe(value, operation['x-resource'], context);
    });
    tool.origin = { kind: 'autotool', spec: spec.info.title, operationId: operation.operationId, digest: digest(spec) };
    return tool;
  });
}
